### PR TITLE
Code Refactoring

### PR DESCRIPTION
**Pull request by sourav15102**
### Pull request title: Refactoring enhancements

I have made some refactoring changes in the code fora few design patterns.
It makes the code more readable and more cohesive.

### Pull request description

- Design pattern: Caching
- Refactoring method: Push down method
- Changes:
> 1. CacheStore contained a lot of methods which can be pushed down to child classes.
> 2. To make it more cohesive, I divided the class into two more classes: CacheStoreRead, CacheStoreWrite.
> 3. This way the code can be considered more extensible.

- Design patter: Commander
- Refactoring method: Decompose conditional
- Changes:
> 1. The if conditional added complexity to the already complex method.
> 2. I took out the if condition and put it in another method, reducing the complexity a little bit.

- Design pattern: Domainmodel
- Refactoring method: Extract class
- Changes:
> 1. There were a lot of responsibilities handled by the class 'Customer'.
> 2. I made a separate class called 'Account' which will handle all the customer's account details.
> 3. Class 'Customer' contains its 'Account'.

- Design pattern: intercepting-filter
- Refactoring method:pull-up method
- Changes:
> 1. All the child classes of 'Abstract Filter' implements execute method, and they are following a pattern.
> 2. We can create a template of this function in 'Abstract Filter', and force child classes to implemented helper method used in the template.

- Design pattern: typeobject
- Refactoring method: extract method
- Changes:
> 1. Some blocks of code in method 'round' in class 'CandyGame' can be taken out.
> 2. Three more functions were created to reduce complexity of function.
> 3. Changes will make it more testable.

- Design pattern: balking
- Refactoring method: Introducing explaining variable
- Changes:
> 1. The variable with value '50' was directly used.
> 2. Introduced variable with explaining names.

### Testing
- There was one test which was failing at the time i forked the repository.
> Design pattern: dirty-flag
- All the other tests are passing.

Please let me know if I need to make any further changes.

### CODE REVIEW

**Review by yoavami — APPROVED** [review]
Changes look good.

File: caching/src/main/java/com/iluwatar/caching/AppManager.java [hunk 1]
```diff
@@ -50,6 +50,8 @@ public class AppManager {
    * Cache Store.
    */
   private final CacheStore cacheStore;
+  private final CacheStoreRead cacheStoreRead;
+  private final CacheStoreWrite cacheStoreWrite;
```
Comment (iluwatar) [inline caching/src/main/java/com/iluwatar/caching/AppManager.java:54]: It's a good inititiative to refactor this, but I find it confusing that we now have three different cache classes and objects.

Comment (iluwatar) [reply caching/src/main/java/com/iluwatar/caching/AppManager.java:54]: One option would be to create a cache class for each of the use cases we want to demonstrate (write-through, write-around, write-behind, cache-aside, read-through)

File: commander/src/main/java/com/iluwatar/commander/Commander.java [hunk 2]
```diff
@@ -245,12 +245,7 @@ private void updateQueue(QueueTask qt) {
       // additional check not needed
       LOG.trace(ORDER_ID + ": Queue time for order over, failed..", qt.order.id);
       return;
-    } else if (qt.taskType.equals(TaskType.PAYMENT) && !qt.order.paid.equals(PaymentStatus.TRYING)
-        || qt.taskType.equals(TaskType.MESSAGING) && (qt.messageType == 1
-        && !qt.order.messageSent.equals(MessageSent.NONE_SENT)
-        || qt.order.messageSent.equals(MessageSent.PAYMENT_FAIL)
-        || qt.order.messageSent.equals(MessageSent.PAYMENT_SUCCESSFUL))
-        || qt.taskType.equals(TaskType.EMPLOYEE_DB) && qt.order.addedToEmployeeHandle) {
+    } else if (this.isTaskDone(qt)) {
```
Comment (iluwatar) [inline commander/src/main/java/com/iluwatar/commander/Commander.java:248]: Maybe `isTaskDone` should be part of `QueueTask` interface?

**Review by iluwatar — CHANGES_REQUESTED** [review]
Overall very nice refactorings. I had a couple of improvement ideas. If you could find time to address them then we can proceed with the merge.

**Pull request closed without merging**